### ISSUE TITLE
jax with iree backend gives compilation error on array update via .at[].set() (works with default cpu backend)

### ISSUE THREAD

**Issue by astrotuna201** [google/jax#12786]
### Description

Trying to use the iree backend with jax yields an error when updating slice with .at[].set()
It looks like the underlying logic does not correctly assemble scatter_indices during broadcast or slice.

Reproducer:

```python
import jax
import jax.numpy as jnp

#crashes later:
jax.config.update('jax_platform_name', 'iree')
#works:
#jax.config.update('jax_platform_name', 'cpu')
#output: [  0   1 -99 -99 -99 -99   6   7]
@jax.jit
def setslice(x):
  myslice = slice(2,-2)
  return x.at[myslice].set(-99)

tst = jnp.arange(8)
result = setslice(tst)
print(result)
```

With the standard "cpu" backend, it yields correct result of 
[  0   1 -99 -99 -99 -99   6   7]

With the "iree" backend (cpu), compilation fails:

```
/usr/bin/python3 simple.py
Traceback (most recent call last):
  File "/Users/aUser/reproducer/simple.py", line 18, in <module>
    result = setslice(tst)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/api.py", line 606, in cache_miss
    out_flat = xla.xla_call(
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/core.py", line 1939, in bind
    return call_bind(self, fun, *args, **params)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/core.py", line 1955, in call_bind
    outs = top_trace.process_call(primitive, fun_, tracers, params)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/core.py", line 701, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/dispatch.py", line 234, in _xla_call_impl
    compiled_fun = xla_callable(fun, device, backend, name, donated_invars,
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/linear_util.py", line 309, in memoized_fun
    ans = call(fun, *args)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/dispatch.py", line 342, in _xla_callable_uncached
    return lower_xla_callable(fun, device, backend, name, donated_invars, False,
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/dispatch.py", line 978, in compile
    self._executable = XlaCompiledComputation.from_xla_computation(
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/dispatch.py", line 1136, in from_xla_computation
    compiled = compile_or_get_cached(backend, xla_computation, options,
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/dispatch.py", line 1054, in compile_or_get_cached
    return backend_compile(backend, serialized_computation, compile_options,
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/profiler.py", line 313, in wrapper
    return func(*args, **kwargs)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/dispatch.py", line 994, in backend_compile
    return backend.compile(built_c, compile_options=options)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/iree.py", line 186, in compile
    iree_binary = iree.compiler.compile_str(
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/core.py", line 278, in compile_str
    result = invoke_immediate(cl, immediate_input=input_bytes)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/binaries.py", line 196, in invoke_immediate
    raise CompilerToolError(process)
jax._src.traceback_util.UnfilteredStackTrace: iree.compiler.tools.binaries.CompilerToolError: Error invoking IREE compiler tool iree-compile
Diagnostics:
/Users/aUser/reproducer/simple.py:15:1: error: 'mhlo.scatter' op expects bounds of the scatter dimensions of updates to be same as the bounds of the corresponding dimensions of scatter indices. For scatter dimension 1, updates bound is 4 , scatter_indices bound is 1.
  return x.at[myslice].set(-99)
^
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/Users/aUser/reproducer/savedir/core-reproducer.mlir`
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/Users/aUser/reproducer/savedir/core-reproducer.mlir`
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/Users/aUser/reproducer/savedir/core-reproducer.mlir`
compilation failed


Invoked with:
 iree-compile /Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-compile - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=llvm-cpu --iree-llvm-embedded-linker-path=/Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false --mlir-pass-pipeline-crash-reproducer=/Users/aUser/reproducer/savedir/core-reproducer.mlir

Need more information? Set IREE_SAVE_TEMPS=/some/dir in your environment to save all artifacts and reproducers.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/aUser/reproducer/simple.py", line 18, in <module>
    result = setslice(tst)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/jax/_src/iree.py", line 186, in compile
    iree_binary = iree.compiler.compile_str(
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/core.py", line 278, in compile_str
    result = invoke_immediate(cl, immediate_input=input_bytes)
  File "/Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/binaries.py", line 196, in invoke_immediate
    raise CompilerToolError(process)
iree.compiler.tools.binaries.CompilerToolError: Error invoking IREE compiler tool iree-compile
Diagnostics:
/Users/aUser/reproducer/simple.py:15:1: error: 'mhlo.scatter' op expects bounds of the scatter dimensions of updates to be same as the bounds of the corresponding dimensions of scatter indices. For scatter dimension 1, updates bound is 4 , scatter_indices bound is 1.
  return x.at[myslice].set(-99)
^
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/Users/aUser/reproducer/savedir/core-reproducer.mlir`
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/Users/aUser/reproducer/savedir/core-reproducer.mlir`
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/Users/aUser/reproducer/savedir/core-reproducer.mlir`
compilation failed


Invoked with:
 iree-compile /Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-compile - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=llvm-cpu --iree-llvm-embedded-linker-path=/Users/aUser/Library/Python/3.9/lib/python/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false --mlir-pass-pipeline-crash-reproducer=/Users/aUser/reproducer/savedir/core-reproducer.mlir

Need more information? Set IREE_SAVE_TEMPS=/some/dir in your environment to save all artifacts and reproducers.
```

Output in reproducer:

```
iree-opt savedir/core-input.mlir 
module @jit_setslice {
  func.func public @main(%arg0: tensor<8xi32>) -> tensor<8xi32> {
    %0 = mhlo.constant dense<2> : tensor<i32>
    %1 = "mhlo.broadcast_in_dim"(%0) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i32>) -> tensor<1xi32>
    %2 = mhlo.constant dense<-99> : tensor<i32>
    %3 = "mhlo.broadcast_in_dim"(%2) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i32>) -> tensor<4xi32>
    %4 = "mhlo.scatter"(%arg0, %1, %3) ({
    ^bb0(%arg1: tensor<i32>, %arg2: tensor<i32>):
      mhlo.return %arg2 : tensor<i32>
    }) {indices_are_sorted = true, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [0], scatter_dims_to_operand_dims = [0]>, unique_indices = true} : (tensor<8xi32>, tensor<1xi32>, tensor<4xi32>) -> tensor<8xi32>
    return %4 : tensor<8xi32>
  }
}
```

Output of core-reproducer:

```
more savedir/core-reproducer.mlir 
#loc0 = loc(unknown)
module @jit_setslice {
  func.func public @main(%arg0: tensor<8xi32> loc(unknown)) -> tensor<8xi32> {
    %0 = mhlo.constant dense<2> : tensor<i32> loc(#loc0)
    %1 = "mhlo.broadcast_in_dim"(%0) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i32>) -> tensor<1xi32> loc(#loc1)
    %2 = mhlo.constant dense<-99> : tensor<i32> loc(#loc0)
    %3 = "mhlo.broadcast_in_dim"(%2) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i32>) -> tensor<4xi32> loc(#loc2)
    %4 = "mhlo.scatter"(%arg0, %1, %3) ({
    ^bb0(%arg1: tensor<i32> loc(unknown), %arg2: tensor<i32> loc(unknown)):
      mhlo.return %arg2 : tensor<i32> loc(#loc3)
    }) {indices_are_sorted = true, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [0], scatter_dims_to_operand_dims = [0]>, unique_indices = true} : (tensor<8xi32>, tensor<1xi32>, tensor<4xi32>) -> tensor<8xi32> loc(#loc3)
    return %4 : tensor<8xi32> loc(#loc0)
  } loc(#loc0)
} loc(#loc0)
#loc1 = loc("jit(setslice)/jit(main)/broadcast_in_dim[shape=(1,) broadcast_dimensions=()]"("/Users/aUser/reproducer/simple.py":15:1))
#loc2 = loc("jit(setslice)/jit(main)/broadcast_in_dim[shape=(4,) broadcast_dimensions=()]"("/Users/aUser/reproducer/simple.py":15:1))
#loc3 = loc("jit(setslice)/jit(main)/scatter[update_consts=() dimension_numbers=ScatterDimensionNumbers(update_window_dims=(0,), inserted_window_dims=(), scatter_dims_to_operand_dims=(0,)) indices_are_sorted=True unique_indices=True mode=GatherScatterMode.FILL_OR_DROP]"("/Users/aUser/reproducer/simple.py":15:1))

{-#
  external_resources: {
    mlir_reproducer: {
      pipeline: "stablehlo-legalize-to-hlo, func.func(mhlo-legalize-control-flow,iree-top-level-scf-to-cfg,iree-mhlo-to-mhlo-preprocessing{order-conv-features=true},canonicalize{  max-iterations=10 region-simplify=true top-down=true},shape-to-shape-lowering), convert-shape-to-std, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true}), inline{default-pipeline=canonicalize{  max-iterations=10 region-simplify=true top-down=true} max-iterations=4 }, iree-util-demote-i64-to-i32, iree-util-demote-f64-to-f32, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,hlo-legalize-shape-computations,iree-mhlo-to-linalg-ext,iree-mhlo-to-linalg-on-tensors), reconcile-unrealized-casts, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true}), iree-mhlo-verify-compiler-input-legality, iree-import-public, iree-import-ml-program, iree-sanitize-module-names, iree-abi-wrap-entry-points, inline{default-pipeline=canonicalize{  max-iterations=10 region-simplify=true top-down=true} max-iterations=4 }, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse), symbol-dce, iree-util-demote-f64-to-f32, func.func(iree-flow-convert-1x1-filter-conv2d-to-matmul,iree-flow-detach-elementwise-from-named-ops,iree-verify-input-legality),util.initializer(iree-flow-convert-1x1-filter-conv2d-to-matmul,iree-flow-detach-elementwise-from-named-ops,iree-verify-input-legality), linalg-named-op-conversion, iree-flow-expand-tensor-shapes, iree-util-fixed-point-iterator{max-iterations=10 pipeline=func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses),iree-util-apply-patterns,iree-util-fold-globals,func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse)}, iree-flow-tensor-pad-to-tensor-insert-slice{skip-one-linalg-use-case=false}, func.func(convert-elementwise-to-linalg,linalg-fold-unit-extent-dims{fold-one-trip-loops-only=false},iree-flow-interchange-generic-ops,resolve-shaped-type-result-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-fusion-of-tensor-ops{fuse-multi-use=false multi-use-fusion-iteration=2},linalg-detensorize{aggressive-mode=false},canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-split-reduction-ops,iree-flow-interchange-generic-ops,iree-flow-dispatch-linalg-on-tensors-pass{aggressive-fusion=false},iree-flow-capture-dispatch-dynamic-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),util.initializer(convert-elementwise-to-linalg,linalg-fold-unit-extent-dims{fold-one-trip-loops-only=false},iree-flow-interchange-generic-ops,resolve-shaped-type-result-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-fusion-of-tensor-ops{fuse-multi-use=false multi-use-fusion-iteration=2},linalg-detensorize{aggressive-mode=false},canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-split-reduction-ops,iree-flow-interchange-generic-ops,iree-flow-dispatch-linalg-on-tensors-pass{aggressive-fusion=false},iree-flow-capture-dispatch-dynamic-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse), iree-flow-initialize-empty-tensors, iree-flow-outline-dispatch-regions, flow.executable(iree-util-strip-debug-ops),func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true}),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true}), iree-flow-deduplicate-executables, flow.executable(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),func.func(iree-flow-cleanup-tensor-shapes,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),util.initializer(iree-flow-cleanup-tensor-shapes,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse), symbol-dce, iree-stream-verify-input, iree-stream-outline-constants, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-stream-conversion, iree-stream-verify-lowering-to-tensors, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-util-combine-initializers, func.func(iree-stream-encode-host-tensors),stream.executable(iree-stream-encode-device-tensors),util.initializer(iree-stream-encode-host-tensors), iree-stream-materialize-builtins, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-uti
```

### What jax/jaxlib version are you using?

jax v0.3.23 jaxlib v0.3.22

### Which accelerator(s) are you using?

CPU

### Additional system info

tested with Python 3.9.6, 3.10 under Linux, macOS

**Comment by hawkinsp** [comment]
Thanks, this is an IREE bug I think. I filed it with IREE as https://github.com/iree-org/iree/issues/10774

**Comment by astrotuna201** [comment]
Seems to be fixed as of IREE [candidate-20221102.315](https://github.com/iree-org/iree/releases/tag/candidate-20221102.315). Many thanks!
